staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -972,19 +1114,31 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer, budget *ti
 	budget.returnUnused(timeout - time.Since(startTime))
 }
 
-// NOTE: sendWatchCacheEvent is assumed to not modify <event> !!!
-func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
+func (c *cacheWatcher) nextBookmarkTime(now time.Time) (time.Time, bool) {
+	// For now we return 2s before deadline (and maybe +infinity is now already passed this time)
+	// but it gives us extensibility for the future(false when deadline is not set).
+	if c.deadline.IsZero() {
+		return c.deadline, false
+	}
+	return c.deadline.Add(-2 * time.Second), true
+}
+
+func (c *cacheWatcher) convertToWatchEvent(event *watchCacheEvent) (watch.Event, bool) {
+	if event.Type == watch.Bookmark {
+		return watch.Event{Type: watch.Bookmark, Object: event.Object.DeepCopyObject()}, true
+	}
+
+	var watchEvent watch.Event
```

Comment:
This means that we perform unnecessary memory allocation, when the watcher is not interested in this object at all.
We can avoid this one if you:
- switch the signature of this function to return *watch.Event
- return nil where this is not needed.

Reply:
watch.Event has two refs only, passing values is more efficient than passing pointers, and this is why we passing value to the channel I guess. I have changed it to *watch.Event.